File: problem1.cpp
# Time Complexity : O(n/k) for all operations, where n is the number of keys and k is the number of buckets.
# Space Complexity : O(n) where n is the number of keys present in the hashset
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# This code uses a has function that finds key % numBuckets and stores the data in the bucket (vector) that is associated to the key value gotten.\
# add(): calls hash function, traverses through the bucket and adds key.
# remove(): calls hash function, traverses through the bucket and removes key.
# add(): calls hash function, traverses through the bucket and checks if key exists.


class MyHashSet {
public:
    int numBuckets;
    vector<vector<int>> buckets;

    int hashFunction(int key)
    {
        return key % numBuckets;
    }

    MyHashSet() {
        numBuckets = 15000;
        buckets = vector<vector<int>>(numBuckets,vector<int>{});
    }
    
    void add(int key) {
        int i = hashFunction(key);
        auto it = find(buckets[i].begin(),buckets[i].end(),key);
        if(it == buckets[i].end())
            buckets[i].push_back(key);
    }
    
    void remove(int key) {
        int i = hashFunction(key);
        auto it = find(buckets[i].begin(),buckets[i].end(),key);
        if(it != buckets[i].end())
            buckets[i].erase(it);
    }
    
    bool contains(int key) {
        int i = hashFunction(key);
        auto it = find(buckets[i].begin(),buckets[i].end(),key);
        if(it != buckets[i].end())
            return true;
        return false;
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: Problem_1.js
// Time Complexity : 
// Add: O(1)
// Remove: O(1)
// Contains: O(1)

// Space Complexity :
// Worst case: O(N) where N is the highest range of numbers that can be added
// Best case: O(sqrt(N))

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

var MyHashSet = function () {
    // Defining size of primary array
    this.buckets = 1000; // This is sqrt of the range of possible numbers given in problem i.e. sqrt(1000000) = 1000
    // Defining size of secondary array
    this.bucketItems = 1000; 
    // This will become a 2d array. Primary array stores the address of each secondary array.
    this.storage = new Array(this.bucket);
};

MyHashSet.prototype.getBucket = function (key) {
    // Get the index of primary array
    return Math.floor(key % this.buckets);
}

MyHashSet.prototype.getBucketItem = function (key) {
    // Get the index of secondary array
    return Math.floor(key / this.bucketItems);
}

/** 
* @param {number} key
* @return {void}
*/
MyHashSet.prototype.add = function (key) {
    let bucket = this.getBucket(key);
    let bucketItem = this.getBucketItem(key);
    // Check if the peimary array is not yet initialized
    if (!this.storage[bucket]) {
        // If the promary array index is 0, we create the secondary array size to bucketItems+1 to cover the edge case of adding number 10^6
        if (bucket == 0) {
            this.storage[bucket] = new Array(this.bucketItems + 1);
        } else {
            this.storage[bucket] = new Array(this.bucketItems);
        }
        // Filling the secondary array with false
        this.storage[bucket].fill(false);
    }
    // Set the double hashed index of element to be added as true
    this.storage[bucket][bucketItem] = true;
};

/** 
* @param {number} key
* @return {void}
*/
MyHashSet.prototype.remove = function (key) {
    let bucket = this.getBucket(key);
    let bucketItem = this.getBucketItem(key);
    // Remove from secondary array only if an address exists in the primary array.
    if (this.storage[bucket]) {
        this.storage[bucket][bucketItem] = false;
    }
};

/** 
* @param {number} key
* @return {boolean}
*/
MyHashSet.prototype.contains = function (key) {
    let bucket = this.getBucket(key);
    let bucketItem = this.getBucketItem(key);
    // Check secondary array only if an address exists in the primary array.
    if (this.storage[bucket]) {
        return this.storage[bucket][bucketItem];
    }
    return false;
};


var obj = new MyHashSet();
obj.add(key);
obj.remove(key);
var param_3 = obj.contains(key);
console.log(param_3);
console.log(obj);

/**
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = new MyHashSet()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */
